internals/overlord/logstate/loki/loki.go

```diff
@@ -0,0 +1,211 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package loki
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/canonical/pebble/internals/overlord/logstate/clienterr"
+	"github.com/canonical/pebble/internals/plan"
+	"github.com/canonical/pebble/internals/servicelog"
+)
+
+const maxRequestSize = 1000
+
+var requestTimeout = 10 * time.Second
+
+type Client struct {
+	target    *plan.LogTarget
+	remoteURL string
+	// buffered entries are "sharded" by service name
+	entries    map[string][]lokiEntry
+	numEntries int
+
+	httpClient http.Client
+	retryAfter *time.Time
+}
+
+func NewClient(target *plan.LogTarget) *Client {
+	return &Client{
+		target:     target,
+		remoteURL:  target.Location,
+		entries:    map[string][]lokiEntry{},
+		httpClient: http.Client{Timeout: requestTimeout},
+	}
+}
+
+func (c *Client) Write(ctx context.Context, entry servicelog.Entry) error {
+	c.entries[entry.Service] = append(c.entries[entry.Service], encodeEntry(entry))
+	c.numEntries++
+
+	if c.numEntries >= maxRequestSize {
+		if c.retryAfter != nil {
+			if time.Now().Compare(*c.retryAfter) == -1 {
```

Comment:
`Time.Compare` was only added in Go 1.20 (mainly for a sort comparison function). It's clearer here to use `Time.After` or `Time.Before`:

```go
if time.Now().Before(*c.retryAfter) {
```

Reply:
thanks, that's what I was looking for